### ISSUE TITLE
Heading Collapse not working if level >1

### ISSUE THREAD

**Issue by JulianSMoore** [jupyterlab/jupyterlab#10714]
## Description

I was not able to find documentation indicating that this unexpected (and to me, undesirable) behaviour is by design

The attached notebook reproduces the unexpected and undesirable behaviour that only H1 headings seem to collapse everything beneath them, all other collapse actions only applying apparently to up to the next heading, whether it is of lower level or not.

Note that the metadata indicates the heading is collapsed.

![image](https://user-images.githubusercontent.com/9352569/127298761-9f551cb7-947a-4d25-a16c-cf4cc6537826.png)

Other info: I have seen (but cannot reproduce) a lower level heading collapsing but then not acquiring a collapsed cell count and not uncollapsing again.

I tried this with two different Python kernels - one with and one without Xeus for debugging; no difference

## Reproduce

[v3.1 Collapse not working.ipynb.zip](https://github.com/jupyterlab/jupyterlab/files/6891804/v3.1.Collapse.not.working.ipynb.zip)

1. Open attached notebook
2. Expand/collapse headings and confirm behaviour is as described therein

## Expected behavior

I expect a heading collapse to collapse and count all cells that are implicit children of the heading by virtue of being lower level headings or other raw/code/markdown cells.

## Context

Note that the Chrome browser in use has NoScript, ABP, Ghostery and UBlock origin plugins, however: ublock origin is disabled for localhost, localhost is trusted by Ghostery, ABP is disabled for localhost, and there was no difference in behaviour between NoScript set to trust localhost and npmjs.org and NoScript disabled for the JupyterLab tab.

- Operating System and version: Win 10, 64-bit Home 21H1
- Browser and version: Chrome Version 91.0.4472.164 (Official Build) (64-bit)
- JupyterLab version: 3.1 pypi_0    pypi
- Other: jupyter-client            6.1.12                   pypi_0    pypi
- Other: jupyter-core              4.7.1                    pypi_0    pypi
- Other: jupyter-server            1.9.0                    pypi_0    pypi
- Other: jupyterlab-pygments       0.1.2                    pypi_0    pypi
- Other: jupyterlab-server         2.6.0                    pypi_0    pypi
- Other: jupyterlab-widgets        1.0.0                    pypi_0    pypi

<details><summary>Troubleshoot Output</summary>
<pre>
None
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
None
</pre>
</details>

<details><summary>Browser Output</summary>

<pre>
None
</pre>
</details>

**Comment by krassowski** [comment]
You seem to think that you have H3 and H4 headings but in fact you don't. What you have is one H1 heading and three H2 headings:

![showing the un-rendered markdown cells with heading levels as in the quote block that follows](https://user-images.githubusercontent.com/5832902/127317668-fadb5549-16c5-4a59-9333-1b1b5bf3e61d.png)

```
# H1 This heading collapses all the way down to the print statement
## H2 This does not collapse at all
## H3 This does not collapse at all
## H4 This collapses the next two cells OK
```

In Markdown the number of hash characters (`#`) determines the level of the heading. The correct number of hash (`#`) characters to achieve what you aim for is demonstrated below:

![showing the un-rendered markdown cells with correct heading levels as in the quote block that follows](https://user-images.githubusercontent.com/5832902/127317876-5efdd2b8-3c55-460f-890b-15d6d6b36422.png)

```
# H1 This heading collapses all the way down to the print statement
## H2 This does not collapse at all
### H3 This does not collapse at all
#### H4 This collapses the next two cells OK
```

You can see on the picture that the levels get a proper indentation in the table of contents now. With proper heading levels it looks to work fine for me:

![works-ok](https://user-images.githubusercontent.com/5832902/127318231-8eb37b69-429f-4964-94eb-11fd09869b52.gif)

**Comment by JulianSMoore** [comment]
@krassowski Apologies about the "MWE" - you are of course right; I screwed up in constructing the "MWE". Unfortunately, there is an issue, I just failed to reproduce it properly. The NB in which I identified the issue has this TOC structure, and collapsing "Function Definitions" - which clearly does have proper structure under it - does nothing.

![image](https://user-images.githubusercontent.com/9352569/127341474-a5e63f0c-8be7-4224-9311-c7036a6d994b.png)

Unfortunately, if I copy and paste Function Definitions structure into a new wb to share as MWE, the problem goes away, so I am sharing share the whole nb with you. If after this there is still not issue for you, then by all means close.

[DESI Photo-Z 43.zip](https://github.com/jupyterlab/jupyterlab/files/6893911/DESI.Photo-Z.43.zip)

**Comment by krassowski** [comment]
I can confirm this is a bug. Could you please try setting:

```
{
    "renderCellOnIdle": false,
    "numberCellsToRenderDirectly": 10000000000000
}
```

in `Advanced Settings Editor` → `Notebook`, saving, reloading JupyterLab and checking if it still happens?

**Comment by echarles** [comment]
Did a quick test with 3.1.rc2 and default settings, and looks to me like it works as expected.

<img width="639" alt="Screenshot 2021-07-28 at 17 29 48" src="https://user-images.githubusercontent.com/226720/127350989-cf69c920-a589-4487-ad5b-385ff9cb2b55.png">

**Comment by krassowski** [comment]
Nah, check out the repro notebook (`DESI Photo-Z 43.ipynb` on `Function Definitions`). It is some king of interaction between cell rendering and collapse feature.

**Comment by krassowski** [comment]
Screencast just in case:

![full-repro](https://user-images.githubusercontent.com/5832902/127352534-433e1dca-9a7f-4280-9f61-df63810962f0.gif)

**Comment by echarles** [comment]
oh, I was using the first reproducer. With the second, indeed, the cells being lazily rendered do not collapse.

**Comment by JulianSMoore** [comment]
> I can confirm this is a bug. Could you please try setting:
> ```
> {
>     "renderCellOnIdle": false,
>     "numberCellsToRenderDirectly": 10000000000000
> }
> ```
> in `Advanced Settings Editor` → `Notebook`, saving, reloading JupyterLab and checking if it still happens?

With those additional settings, Function Definitions collapses and shows a button with "+89 cells hidden" - further good news: previously (but not reported) clicking the "cells hidden" button where it did appear also did not unhide the hidden cells (the only way to make them reappear was to click on a hidden heading in the ToC (heading visible in ToC, hidden in nb itself).

So, that at least mitigates the reported issue. Is it in fact also the fix or is there still a bug to be resolved?

**Comment by willemavjc** [comment]
Can confirm @krassowski 's settings make things work as expected. I'm currently working on a (huge?) notebook with 50 or 60 headers going from level 1 to 4. (Un)Collapsing was stopping after the few first ones. The most scaring part was that in fact it was acting like it could not do the action but closing and reopening the notebook was resulting in cells disappearing! (Like they never even existed!)

**Comment by krassowski** [comment]
Closing as fixed by #10793 which was backported in #10797 and released with JupyterLab 3.1.6.
